notebooks/MixedModelsTutorial_basics.jl

```diff
@@ -66,124 +71,207 @@ Number of scores: 525126
 
-We extract a random sample of 1000 children from the Sex (2)  x Test (5) cells of the design. Cohort and School are random. 
+For the prupose of the tutorial, we extract a random sample of 1000 boys and 1000 girls. `Child`, `School`, and `Cohort` are grouping variables. Traditionally, they are called random factors because the units (levels) of the factor are assumed to be a random sample from the population of their units (levels).  
+
+Cohort has only nine "groups" and could have been included as a set of polynomical fixed-effect contrasts rather than a random factor. This choice warrants a short excursion: The secular trends are very different for different tests and require the inclusion of interaction terms with `Test` contrasts (see Figure 4 in [Fühner et al. (2021)](https://rdcu.be/cwSeR). The authors opted to absorb these effects in cohort-related variance components for the `Test` contrasts and plan to address the details of secular changes in a separate analysis.
+
+For complex designs, when they are in the theoretical focus of an article, factors and covariates should be specified as part of the fixed effects. If they are not in the theoretical focus, but serve as statistical control variables, they could be put in the RES - if supported by the data.
+"""
+
+# ╔═╡ 92038a7c-69f6-445d-ba29-03ae4cf9f115
+md"**Unstack:** Data are in the long format. We need them in a child-based wide format -- one row per child"
+
+# ╔═╡ fe027497-7ed1-4780-871b-547d4f466072
+df_w = unstack(dfx, :Test, :zScore)
+
+# ╔═╡ 29d0a540-29ce-43de-9dfe-75e535ff94eb
+md""" **Stratified sampling:** `MersenneTwister(42)` specifies `42` as the seed for the random number generator to ensure reproducibility of the stratification. The stratified sample is stored in data frame `dat`.
 """
 
-# ╔═╡ 6e3adecf-98ff-4437-8e16-3854381ca38b
+# ╔═╡ b3f32059-0d7b-4b66-8559-7c8a9f4d14d3
+
 begin
-    dat =
-	@chain df begin
-  	  @transform(:Sex = :Sex == "female" ? "Girls" : "Boys")
-   	  @groupby(:Test, :Sex)
-   	  combine(x -> x[sample(1:nrow(x), 1000), :])
+dat_w =
+	@chain df_w begin
+   	  @groupby(:Sex)
+   	  combine(x -> x[sample(MersenneTwister(42), 1:nrow(x), 1000), :])
 	end
 end
 
-# ╔═╡ 51b77ca3-374d-4701-b936-b4a40834cf70
-md"""
-#### 1.2.4 Transformations
-"""
 
-# ╔═╡ 7440c439-9d4c-494f-9ac6-18c9fb2fe144
+# ╔═╡ 1f71f182-d840-43fd-98e5-721909e67b60
+md" **Stack and configure data:** Now convert back from wide to long format -- one row per observation. We delete 2% rows with missing data." 
+
+# ╔═╡ cf28cee0-7a33-4910-988a-d60aba25522e
 begin
-	transform!(dat, :age, :age => (x -> x .- 8.5) => :a1); # centered age (linear)
-	select!(groupby(dat,  :Test), :, :score => zscore => :zScore); # z-score
+	dat = stack(dat_w, 7:11);
+	rename!(dat, :variable => :Test, :value => :zScore);
+	dropmissing!(dat, :zScore)
+	dat.Test = categorical(dat.Test);
+	levels!(dat.Test, ["Endurance", "Coordination", "Speed", 
+			           "PowerLOW", "PowerUP"])
```

Comment:
I am not clear on the need to store the `Test` factor with the abbreviated names and then reset them.  If you want those names and that particular order, wouldn't it be easier to store the factor with those names and order in the Arrow file?

Reply:
If you want to sample 1000 Girls and 1000 Boys, wouldn't it be easier to sample the `:Child` identifiers and then use `subset` or `@subset` to get the sample in the long form.  The missing data are introduced by the `unstack/stack` sequence.

Reply:
I would probably go about this by creating a `Student` table as
```jl
Student = unique(select(df, :Cohort, :School, :Child, :Sex, :age))
```
and doing the stratified sample there
```jl
 samp = let rng = MersenneTwister(42)
     combine(groupby(Student, :Sex), x -> x[rand(rng, 1:nrow(x), 1000), :])
end
```
Then write a function to check whether a `Child` identifier is in the sample
```jl
insamp(x) = x ∈ samp.Child
```
and use that in a call to `@subset`
```jl
@subset(df, insamp(:Child))
```

Reply:
Really nice. Thank you!